### DISCUSSION TITLE
Multicore sort when database is stored on disk?

### DISCUSSION THREAD

**Discussion by vedantroy** [duckdb/duckdb#5749] [Q&A] [answered]
I have a 545GB database compromised of a few thousand parquet files. Each parquet file has 2 columns: name (string), and data (bytes).

Initially, I loaded the entire database into an in-memory duckdb instance. This worked fine.
Also, when I sorted the resulting table--all of the cores on my 96 core machine were being used.
Alas, even though I had 1TB of memory, I still received an OOM error from duckdb.

On my 2nd attempt, I made the database an on disk database before loading in the parquet files.
This also worked. But, now when I do the sorting, I can see that only a single core on my computer is being used.
From looking at the rate of progress, it seems like it will be at least a few hours before all 20 million items get sorted.
On the bright side, memory usage is flat at ~ 570GB, so I don't think I'll OOM.

Is there anything I can do to speed up sorting without OOM-ing?

One node is that my name column is actually like this:
```
hashc_2
hasha_1
hashb_2
hashc_1
hashb_1
hasha_2
```

and I'm trying to sort it in: hash order, and then index order. Even though, I could forgo the index order if necessary.
Maybe by converting the names from strings to integer columns, I can get a speedup?

**Comment by Alex-Monahan** [comment, marked as answer]
When using an in memory DB, be sure to specify a temp directory so that DuckDB can spill to disk if needed! That should allow you to get the speed of in memory, but fall back to disk if needed.

Also, that is some serious RAM! 😀 Mind sharing a bit about your use case? Just curious!

**Reply by vedantroy** [reply]
I'm sorting a machine learning dataset composed of a large number of pairs of (file name, binary file).

**Comment by vedantroy** [comment]
@Alex-Monahan Thanks for the response! I'm guessing the right command is: `PRAGMA temp_directory='/path/to/temp.tmp'` (which seems to specify either a file or directory? unclear.)

**Reply by Alex-Monahan** [reply]
Looks like a file is what you want based on this test! I agree the name is not very specific...
https://github.com/duckdb/duckdb/blob/f411bd114705cc79d42e43fd4cfb10c3be5b0ddb/test/sql/storage/buffer_manager_temp_dir.test

**Reply by Alex-Monahan** [reply]
Let us know if that does the trick!
